ESQL: Remove redundant `TopN`

kanoshiou opened the pull request:
Redundant `TopN` operations were not being removed from the execution plan, leading to unnecessary performance overhead.

We can safely consider a child `TopN` plan redundant if it shares the same sort order as its parent. There's no need to compare their limit values, since `PushDownAndCombineLimits` already pushes down the lower limit value, making it safe to remove the child `TopN`.

Also, the `PruneRedundantOrderBy` rule was only executed before the `ReplaceLimitAndSortAsTopN`, so it could only detect redundant `OrderBy` operations but not redundant `TopN` operations that were created later in the optimization pipeline.

Closes #131233

kanoshiou committed:
Remove redundant `TopN`

kanoshiou committed:
Update docs/changelog/131833.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

kanoshiou committed:
Merge branch 'main' into remove-redundant-topn

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

smalyshev commented:
@elasticsearchmachine test this please

kanoshiou committed:
Merge branch 'main' into remove-redundant-topn

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java:
```diff
@@ -58,25 +58,32 @@ protected LogicalPlan rule(LogicalPlan plan) {
+                    result.add(childTopN);
+                    toCheck.push(childTopN.child());
+                }
             } else if (p instanceof SortAgnostic) {
```
The new version of the rule is incorrect, I'm afraid.

We continue going down the plan as long as the children are `SortAgnostic`. However, being `SortAgnostic` doesn't automatically imply that the plans behave the same when we feed them an unlimited number of rows.

For instance, `INLINE STATS` is `SortAgnostic` (resp. `InlineJoin` is). But removing a `TopN` from before an `INLINE STATS` changes the meaning of the query.

Also, queries can become much more expensive: For instance, `MvExpand` is `SortAgnostic`, but `SORT a | LIMIT 10 | MV_EXPAND b | SORT a | LIMIT 10` is generally cheaper to run than just `MV_EXPAND b | SORT a | LIMIT 3`.

alex-spies reviewed (changes requested):
Thanks @kanoshiou and apologies that it took us so long to get back to you! Your contributions are very much appreciated!

Unfortunately, I think the new version of `PruneRedundantOrderBy` is affecting too many queries, including ones where removing an upstream TopN is incorrect. I give examples below.

I think we should limit the scope by adding a new rule that only looks for subsequent TopNs, specifically, and place that rule into the `cleanup` batch of the `LogicalPlanOptimizer`. That's more obviously correct. Cases when there are query plan nodes between two `TopN`s and the upstream one should be pruned are trickier to handle and likely require to mark query plans in some way when they are compatible with this optimization.

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

kanoshiou committed:
PruneRedundantTopN

kanoshiou commented:
Thank you for your review @alex-spies! I've made several changes based on your feedback. However, I'm still finding it challenging to handle cases where there are query plan nodes between two `TopN`. Any guidance on how to approach these scenarios would be greatly appreciated.

From the examples you provided in your comment, removing the second `TopN` in such cases could introduce performance issues (except for nodes like `DROP` and `KEEP`, perhaps?) or even alter query semantics. In your experience, under what conditions can we safely remove the redundant `TopN` without compromising correctness or efficiency?

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

kanoshiou committed:
Update

kanoshiou committed:
Add more tests

kanoshiou committed:
Merge branch 'main' into remove-redundant-topn

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

alex-spies reviewed (changes requested):
Thanks a lot for your effort @kanoshiou ! This one's not easy!

I still think we should limit the scope to `... | TOP N | TOP M | ...` . It's very difficult to get the rule right if there are other nodes between the `TOP N` and `TOP M`. We have something similar in `HoistRemoteEnrichTopN`, but I don't think this can be applied here 1:1. In fact, the complexity increased sine `HoistRemoteEnrichTopN` was added because now TOP Ns can also be `local`, meaning, they don't necessitate breaking the pipeline. (And getting `HoistRemoteEnrichTopN` was highly complex, too.)

kanoshiou committed:
Merge branch 'main' into remove-redundant-topn

kanoshiou committed:
refactor(esql): Simplify PruneRedundantTopN optimization logic

- Refactor PruneRedundantTopN to use a more straightforward approach for removing redundant TopN operations
- Replace complex breadth-first recursive search with a simpler while loop traversal
- Remove unnecessary method for checking removable child nodes
- Improve performance and readability of TopN pruning logic
- Update test case to reflect simplified optimization approach

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/topN.csv-spec

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Update test

kanoshiou committed:
Merge branch 'refs/heads/main' into remove-redundant-topn

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/topN.csv-spec

wchaparro commented:
<!-- analytics-pr-review-management:stall-warn -->

Hi @kanoshiou — this PR has been identified as **stalled** (open and without meaningful updates for 30+ days).

Please address the requested changes when you can.

It will be **closed in two weeks** if there is no further activity. Please reply, push an update, or close it yourself if it is no longer needed.

— Analytics PR review management

kanoshiou committed:
Merge branch 'main' into remove-redundant-topn

alex-spies commented:
Heya @wchaparro , this one is also pending on us - kanoshiou already addressed comments as far as I could see. It's been in my to-re-review list but high prio stuff came up so I didn't manage to get back to this. It is still relevant.

And sorry @kanoshiou for letting you wait so long!
